### PR TITLE
New lint: `unnecessary_map_or`

### PR DESCRIPTION
**Pull request by Jacherr**
Closes https://github.com/rust-lang/rust-clippy/issues/10118

This lint checks `map_or` method calls to check if they can be consolidated down to something simpler and/or more readable.

For example, the code 
```rs
let x = Some(5);
x.map_or(false, |n| n == 5)
```

can be rewritten as 
```rs
let x = Some(5);
x == Some(5)
```

In addition, when the closure is more complex, the code can be altered from, say,
```rs
let x = Ok::<Vec<i32>, i32>(vec![5]);
x.map_or(false, |n| n == [5])
```
into
```rs
let x = Ok::<Vec<i32>, i32>(vec![5]);
x.is_some_and(|n| n == [5])
```

This lint also considers cases where the `map_or` can be chained with other method calls, and accommodates accordingly by adding extra parentheses as needed to the suggestion.

changelog: add new lint `unnecessary_map_or`

### CODE REVIEW

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:85]: This should be either `snippet_with_applicability` or `snippet_opt`

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:76]: I don't really like checking this again. imo you should create an enum based on the results of the first check - i.e., `Result`/`Option`, then give it a method like `sugg`. As an example take a look at `unwrap_used`: https://github.com/Centri3/rust-clippy/blob/71c54137eadc730d01fe15c17df8bf3ba5d3b04b/clippy_lints/src/methods/unwrap_expect_used.rs#L12-L15

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:49]: Same here, maybe `method_name`?

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:62]: Could we use `maybe_par` with `Sugg` instead?

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:64]: nit: Please inline some of these, like `"{comparator}"`.

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:41]: Can you add a comment here explaining that this allows linting both `y == x` and `x == y`? I also think "binding" here is misleading, this should be `param` instead

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:54]: We can use `as_str` here

**Comment by WaffleLapkin** [comment]
@Jacherr this seems to be waiting on you, since Catherine left you a review.
@rustbot author

**Comment by Jacherr** [comment]
Hi, apologies for the lack of movement on this. I’ll make changes based on the review now, and once everything is up to standard I’ll go through the conflicts.

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Jacherr) [reply clippy_lints/src/methods/unnecessary_map_or.rs:87]: Could you clarify exactly what you mean here? Would something like `make_binop(BinOpKind::Eq, Sugg::hir(recv), Sugg::hir(non_binding_location)).maybe_par()` be what you're looking for?

**Comment by xFrednet** [comment]
Hey @Jacherr Sorry for the delay. @Centri3 was sadly busy, but now she indicated that she'll have time for reviews again. Would you mind rebasing again? Next time there is a long delay, you can also pick a new reviewer with `r? clippy`.

**Comment by Jacherr** [comment]
I did rebase but it's not showing up in the PR for some reason 😕

File: tests/ui/unnecessary_map_or.rs
Comment (Centri3) [inline tests/ui/unnecessary_map_or.rs:1]: Just needs tests for external and procedural macros

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:109]: We should call `is_from_proc_macro` here

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:104]:
```suggestion
            let suggested_method = variant.method_name();
            (format!("{recv_callsite}.{sugg}({span_callsite})",), suggested_method)
```
Since this has the same name as the `sugg` above, which is the other value in the tuple

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [reply clippy_lints/src/methods/unnecessary_map_or.rs:87]: Basically, checking precedence here I think is so the code doesn't change its meaning. Using `maybe_par` will ensure (maybe while requiring an extra run of `check`) that the precedence isn't changed.

Yeah, that snippet actually looks perfect

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:76]: Small formatting change, surprised rustfmt didn't catch this
```suggestion
            && let closure_body = cx.tcx.hir().body(map_closure.body)
            && let closure_body_value = closure_body.value.peel_blocks()
            && let ExprKind::Binary(op, l, r) = closure_body_value.kind
            && let Some(param) = closure_body.params.first()
            && let PatKind::Binding(_, hir_id, _, _) = param.pat.kind
            // checking that map_or is one of the following:
            // .map_or(false, |x| x == y)
            // .map_or(false, |x| y == x) - swapped comparison
            // .map_or(true, |x| x != y)
            // .map_or(true, |x| y != x) - swapped comparison
            && ((BinOpKind::Eq == op.node && !def_bool) || (BinOpKind::Ne == op.node && def_bool))
            && let non_binding_location = if path_to_local_id(l, hir_id) { r } else { l }
            && switch_to_eager_eval(cx, non_binding_location)
            // xor, because if its both then thats a strange edge case and
            // we can just ignore it, since by default clippy will error on this
            && (path_to_local_id(l, hir_id) ^ path_to_local_id(r, hir_id))
            && !is_local_used(cx, non_binding_location, hir_id)
            && typeck_results.expr_ty(l) == typeck_results.expr_ty(r)
```

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:46]: Would prefer using a bunch of `let..else`s over this

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:114]:
```suggestion
            format!("this `map_or` is redundant"),
            "use {method} instead",
```

**Review by Centri3 — APPROVED** [review]
Just a couple more changes then this is good <3

**Comment by xFrednet** [comment]
Hey, this is a ping from triage. @Centri3 can you give this PR a review? It's totally fine if you don't have the time right now, you can reassign the PR to a random team member using `r? clippy`.

@rustbot ready

**Comment by xFrednet** [comment]
Hey, this is a ping from triage. @Centri3  can you give this PR a review? It's totally fine if you don't have the time right now, you can reassign the PR to a random team member using `r? clippy`.

@rustbot ready

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:45]: Should be the final check before linting since this is rather expensive

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:66]: We can refactor this to instead call `type_diagnostic_name` in a match

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:97]: This can still use `Sugg::maybe_par`

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:107]:
```suggestion
            "a standard comparison",
```

File: clippy_lints/src/methods/unnecessary_map_or.rs
Comment (Centri3) [inline clippy_lints/src/methods/unnecessary_map_or.rs:130]: I just noticed that this might not be machine applicable. Don't we need to check for `PartialEq` somewhere if we'll change it to `opt == <something>`?

Comment (samueltardieu) [reply clippy_lints/src/methods/unnecessary_map_or.rs:130]: Yes. This is not a problem for an `Option<T>` as we know that `T` already implements `PartialEq` if the original code was `x.map_or(false, |x| == value)`, and thus `Option<T>` implements `PartialEq` as well. However, for `Result<T, E>`, `E` must implement `PartialEq` as well, or the comparison cannot be done at the `Result<T, E>` level.

For example, the following code will lead to an incorrect suggestion:

```rust
fn main() {
    struct S;
    let r: Result<i32, S> = Ok(3);
    let _ = r.map_or(false, |x| x == 7);
}
```

because the suggested replacement `r == Ok(7)` won't work.

**Comment by Jacherr** [comment]
@rustbot ready

Apologies for the slow turnover again. I hope this is the last of the issues now and this can finally be merged 😅

File: tests/ui/unnecessary_map_or.fixed
Comment (samueltardieu) [inline tests/ui/unnecessary_map_or.fixed:12]: These `(` should be removed, shouldn't they?

Comment (Jacherr) [reply tests/ui/unnecessary_map_or.fixed:12]: I think this is just an effect of `maybe_par` deciding they should be present. While I'd usually agree, I'm not sure there is a better way to tackle that?

Comment (Centri3) [reply tests/ui/unnecessary_map_or.fixed:12]: `maybe_par` is rather overeager, but `unused_parens` will usually catch them anyway. New lints triggering on fixed code in a cascade is fine

**Comment by samueltardieu** [comment]
Is it time to open a FCP for this lint? I keep seeing `.map_or(false, …)` used everywhere.

**Comment by Jacherr** [comment]
Squash done, ready for final comments/changes then hopefully can finally get this merged. Apologies again for how long this took.

**Comment by samueltardieu** [comment]
@Jacherr It looks like `.map_or(true, …)` can be linted as well, using `.is_none_or(…)`. I played with it a bit but I have to go AFK right now, so feel free to integrate https://github.com/samueltardieu/rust-clippy/commit/a468081648451c93b1ae334e17923d7b062c2a90 in your code if you wish, and update all the lints which trigger this code. Otherwise, I'll submit a complete PR after yours is merged.

Note that tests that don't involve `PartialEq` are machine applicable as far as I can see, so I've changed this as well, but not looked closely at why the others aren't.

**Comment by Centri3** [comment]
I am happy with merging this as-is if no concerns come up, and having that done later.

**Comment by samueltardieu** [comment]
> I am happy with merging this as-is if no concerns come up, and having that done later.

This seems reasonable indeed.

**Comment by samueltardieu** [comment]
@Jacherr Commit https://github.com/rust-lang/rust-clippy/pull/13653/commits/430235e80d022faefb866b1e8e193e1d49f1501f is an updated version of your commit compatible with the current `master`

**Comment by samueltardieu** [comment]
(an alternative is to merge https://github.com/rust-lang/rust-clippy/pull/13653 which contains this PR, updated, and the `.map_or(true, …)` case as well)

**Comment by Jacherr** [comment]
Hopefully, everything is in order for merge now. Ideally, as mentioned in the FCP, this can be merged and the additional `map_or(true` case added in post. It would be nice to put this PR to bed properly given how long it's been going back and forth.

**Review by samueltardieu — APPROVED** [review]

**Comment by Centri3** [comment]
Thanks for your work on this :) I doubt anybody will object the lint being added so let's put it to rest :heart:

**Pull request merged**
